metrics-operator/controllers/analysis/workersPool.go

```diff
@@ -0,0 +1,190 @@
+	}
+
+	for w := 1; w <= aw.numWorkers; w++ {
+		go aw.RetrieveProvider(ctx, w)
+	}
+}
+
+func (aw WorkersPool) CollectAnalysisResults() map[string]metricstypes.ProviderResult {
+
+	results := make(map[string]metricstypes.ProviderResult, aw.numJobs)
+	for a := 1; a <= aw.numJobs; a++ {
+		res := <-aw.results
+		aw.Log.Info("collected result")
+		// Making sure error gets propagated
+		results[analysis.ComputeKey(res.Objective)] = res
+
+	}
+	aw.stopProviders()
+	close(aw.results)
+	return results
+}
+
+func (aw WorkersPool) stopProviders() {
+	for _, ch := range aw.providers {
+		close(ch)
+	}
+}
+
+func generateQuery(query string, selectors map[string]string) (string, error) {
+	tmpl, err := template.New("").Parse(query)
+	if err != nil {
+		return "", fmt.Errorf("could not create a template: %v", err)
+	}
+
+	var resultBuf bytes.Buffer
+	err = tmpl.Execute(&resultBuf, selectors)
+	if err != nil {
+		return "", fmt.Errorf("could not template the args: %v", err)
+	}
+
+	return resultBuf.String(), nil
+}
+
+func (aw WorkersPool) RetrieveProvider(ctx context.Context, id int) {
+
+	for _, j := range aw.Objectives[id] {
+
+		aw.Log.Info("worker", "id:", id, "started  job:", j.AnalysisValueTemplateRef.Name)
+
+		template := &metricsapi.AnalysisValueTemplate{}
+		if j.AnalysisValueTemplateRef.Namespace == "" {
+			j.AnalysisValueTemplateRef.Namespace = aw.Namespace
+		}
+		err := aw.Client.Get(ctx,
+			types.NamespacedName{
+				Name:      j.AnalysisValueTemplateRef.Name,
+				Namespace: j.AnalysisValueTemplateRef.Namespace},
+			template,
+		)
+
+		if err != nil {
+			aw.Log.Error(err, "Failed to get the correct Provider")
+			aw.results <- metricstypes.ProviderResult{Objective: j.AnalysisValueTemplateRef, Err: err}
+			continue
+		}
+
+		providerRef := &metricsapi.KeptnMetricsProvider{}
+		if template.Spec.Provider.Namespace == "" {
+			template.Spec.Provider.Namespace = aw.Namespace
+		}
+		err = aw.Client.Get(ctx,
+			types.NamespacedName{
+				Name:      template.Spec.Provider.Name,
+				Namespace: template.Spec.Provider.Namespace},
+			providerRef,
+		)
+
+		if err != nil {
+			aw.Log.Error(err, "Failed to get Provider")
+			aw.results <- metricstypes.ProviderResult{Objective: j.AnalysisValueTemplateRef, Err: err}
+			continue
+		}
+
+		templatedQuery, err := generateQuery(template.Spec.Query, aw.Analysis.Spec.Args)
+		if err != nil {
+			aw.Log.Error(err, "Failed to substitute args in template")
+			aw.results <- metricstypes.ProviderResult{Objective: j.AnalysisValueTemplateRef, Err: err}
+			continue
+		}
+		//send job to provider solver
+		aw.providers[providerRef.Spec.Type] <- metricstypes.ProviderRequest{
+			Objective: &j,
+			Query:     templatedQuery,
+			Provider:  providerRef,
+		}
+	}
+}
+
+// TODO add timeout and spec save of unfinished analysis
+func (aw WorkersPool) Evaluate(ctx context.Context, providerType string, obj chan metricstypes.ProviderRequest) {
+	provider, err := providers.NewProvider(providerType, aw.Log, aw.Client)
```

Comment:
we might want to decouple by injecting something like a provider factory into this component, to be able to use mocked provider implementations in the unit tests. given the size of this PR, I would also be fine with this being done in a follow up refactoring

Reply:
yes initially I had also the provider retrieval injected to simplify test, I would probably have both :(